subprojects/core/src/test/groovy/org/gradle/internal/operations/DefaultBuildOperationQueueTest.groovy

```diff
@@ -171,28 +226,31 @@ class DefaultBuildOperationQueueTest extends Specification {
         lease.leaseFinish() // Release worker lease to allow operation to run, when there is max 1 worker thread
 
         when:
-        runs.times { operationQueue.add(new SynchronizedBuildOperation(operationAction, startedLatch, releaseLatch)) }
+        def waitForCompletionThread = new Thread({
+            workerRegistry.runAsWorkerThread {
+                runs.times { operationQueue.add(new SynchronizedBuildOperation(operationAction, startedLatch, releaseLatch)) }
+                operationQueue.waitForCompletion()
+            }
+        })
+        waitForCompletionThread.start()
+
+        and:
         // wait for operations to begin running
-        startedLatch.await()
+        startedLatch.await(30, TimeUnit.SECONDS)
 
         and:
         operationQueue.cancel()
 
         and:
         // release the running operations to complete
         releaseLatch.countDown()
-        workerRegistry.runAsWorkerThread {
-            operationQueue.waitForCompletion()
-        }
+        waitForCompletionThread.join(30000)
 
         then:
         expectedInvocations * operationAction.run()
 
         where:
         runs | threads
-        0    | 1
-        0    | 4
-        0    | 10
```

Comment:
Running this with 0 operations doesn't make sense, since this test is about ensuring unstarted operations don't start after the queue is cancelled.